### PR TITLE
Add coverage library

### PR DESCRIPTION
**Pull request by lynzrand**
This PR is to facilitate PR 2784 in the compilers repo to add code coverage support

### CODE REVIEW

**Comment by Lampese** [comment]
Can you complete the CLA first? Don't forget to log in to it after signing, and cla-check will work fine the next time you commit to rerunning the action.

**Comment by lynzrand** [comment]
Already did, will trigger the pipeline by force pushing some time later. BTW I'm that intern working on coverage the last few days.

File: coverage/coverage.mbt [hunk 1]
```diff
@@ -0,0 +1,100 @@
+pub struct CoverageCounter {
+  priv counter : Array[Int]
+}
+
```
Comment (bobzhang) [inline coverage/coverage.mbt:4]: type defaults to abstract, so you don't need annotate anything here

File: coverage/coverage.mbt [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+fn CoverageCounter::incr(self : CoverageCounter, idx : Int) -> Unit {
+  let counter = self.counter[idx];
+  if counter < 0x7fffffff { // prevent overflow
+    self.counter[idx] = counter + 1
+  }
+}
+
+fn CoverageCounter::to_string(self : CoverageCounter) -> String {
```
Comment (bobzhang) [inline coverage/coverage.mbt:16]: is this for debugging? if so, derive would be enough

File: coverage/coverage.mbt [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    abort("error")
+  }
+}
+
+test "incr_counter" {
+  let counter = CoverageCounter::new(10)
+  counter.incr(0)
+  counter.incr(9)
+  let result = counter.to_string()
+  if result != "[1, 0, 0, 0, 0, 0, 0, 0, 0, 1]" {
```
Comment (bobzhang) [inline coverage/coverage.mbt:44]: using assertion package

File: coverage/coverage.mbt [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    println(result)
+    abort("error")
+  }
+}
+
+let counters : Ref[List[(String, CoverageCounter)]] = { val: Nil }
+
+fn escape_string(s : String) -> String {
+  let buf = Buffer::make(s.length())
+  let mut ix = 0
```
Comment (bobzhang) [inline coverage/coverage.mbt:54]: where is this used, this could be moved into String package?

File: coverage/coverage.mbt [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+      }
+    }
+    ix += 1
+  }
+  buf.to_string()
+}
+
+pub fn track(name : String, counter : CoverageCounter) -> Unit {
+  counters.val = Cons((name, counter), counters.val)
+}
```
Comment (bobzhang) [inline coverage/coverage.mbt:77]: CoverageCounter.track -- export as method?

File: coverage/coverage.mbt (hunk 2 above)
Comment (lynzrand) [reply coverage/coverage.mbt:16]: It's for exporting the coverage data in `@coverage.end()`, so it must be in JSON format. I don't know if deriving has such guarantees.

File: coverage/coverage.mbt (hunk 4 above)
Comment (lynzrand) [reply coverage/coverage.mbt:54]: It's for escaping the coverage data's key (filename). I'll do an overhaul and put it into Strings

File: coverage/coverage.mbt (hunk 2 above)
Comment (lynzrand) [reply coverage/coverage.mbt]: Update: converted it to a custom implementation of `Debug` and derived `Show`

File: string/string.mbt [hunk 6]
```diff
@@ -11,3 +11,69 @@
...
+      _ =>
+        if ch.to_int() < 0x20 {
+          buf.write_char('\\')
+          buf.write_char('x')
+          let ich = ch.to_int()
+          buf.write_char(to_hex_digit(ich / 16))
+          buf.write_char(to_hex_digit(ich % 16))
+        } else {
+          buf.write_char(ch)
+        }
```
Comment (lynzrand) [inline string/string.mbt:52]: Do we need to factor this part out to `char` as `Char::escape_to`? Also, do we need to additionally provide escaping for non-ascii chars?

File: coverage/coverage.mbt [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+/// }
+/// ----- END MOONBIT COVERAGE -----
+/// ```
+pub fn end() -> Unit {
+  println("----- BEGIN MOONBIT COVERAGE -----")
+  println("{")
+  let mut ix = 0
+  loop counters.val {
+    Cons((name, counter), xs) => {
+      if ix != 0 {
```
Comment (bobzhang) [inline coverage/coverage.mbt:77]: rename this to a separate function and write a simple test?

**Review by bobzhang — COMMENTED** [review]
after second thoughts, the added String::escape is only useful for coverage, we can keep it internal to such package.
The PR is good to go modulo my last two comments

File: coverage/coverage.mbt (hunk 7 above)
Comment (lynzrand) [reply coverage/coverage.mbt:76]: Good idea, will do

Comment (bobzhang) [reply coverage/coverage.mbt:76]: note we can use a trait object so that it can be output to a string or console

**Comment by lynzrand** [comment]
> after second thoughts, the added String::escape is only useful for coverage, we can keep it internal to such package. The PR is good to go modulo my last two comments

Maybe it's useful for debug outputting strings?

File: coverage/coverage.mbt (hunk 7 above)
Comment (lynzrand) [reply coverage/coverage.mbt:76]: > a trait object

AFAIK we don't have such infrastructure (like `trait Write`) yet, should I add one in another PR?

Comment (bobzhang) [reply coverage/coverage.mbt]: I added a new commit, check the TODO in the comment

**Review by bobzhang — APPROVED** [review]

**Pull request merged**